bpf/conntrack/cleanup.go

```diff
@@ -47,96 +48,140 @@ func DefaultTimeouts() Timeouts {
 	}
 }
 
-type LivenessScanner struct {
-	timeouts Timeouts
+// ScanVerdict represents the set of values returned by EntryScan
+type ScanVerdict int
+
+const (
+	// ScanVerdictOK means entry is fine and should remain
+	ScanVerdictOK = iota
+	// ScanVerdictDelete meand entry should be deleted
+	ScanVerdictDelete
+)
+
+// EntryGet is a function prototype provided to EntryScanner in case it needs to
+// evaluate other entries to make a verdict
+type EntryGet func(Key) (Value, error)
+
+// EntryScanner is a function prototype to be called on every entry by the scanner
+type EntryScanner func(Key, Value, EntryGet) ScanVerdict
+
+// Scanner iterates over a provided conntrack map and call a set of EntryScanner
+// functions on each  entry in the order as they were passed to NewScanner. If
+// any of the EntryScanner returns ScanVerdictDelete, it deletes the entry, does
+// not call any other EntryScanner and continues the iteration.
+//
+// It provides a delete sae iteration over the conntrack table for multiple
+// evaluation functions, to keep their implementation simpler.
+type Scanner struct {
 	ctMap    bpf.Map
-	dsr      bool
-	NowNanos func() int64
+	scanners []EntryScanner
 }
 
-func NewLivenessScanner(timeouts Timeouts, dsr bool, ctMap bpf.Map) *LivenessScanner {
-	return &LivenessScanner{
-		timeouts: timeouts,
+// NewScanner returns a scanner for the given conntrack map and the set of
+// EntryScanner. They are executed in the provided order on each entry.
+func NewScanner(ctMap bpf.Map, scanners ...EntryScanner) *Scanner {
+	return &Scanner{
 		ctMap:    ctMap,
-		dsr:      dsr,
-		NowNanos: bpf.KTimeNanos,
+		scanners: scanners,
 	}
 }
 
-func (l *LivenessScanner) Scan() {
-	err := l.ctMap.Iter(func(k, v []byte) {
-		ctKey := keyFromBytes(k)
-		ctVal := entryFromBytes(v)
+// Scan executes a scanning iteration
+func (s *Scanner) Scan() {
+	err := s.ctMap.Iter(func(k, v []byte) {
+		ctKey := KeyFromBytes(k)
+		ctVal := ValueFromBytes(v)
 		log.WithFields(log.Fields{
 			"key":   ctKey,
 			"entry": ctVal,
 		}).Debug("Examining conntrack entry")
```

Comment:
This could be a fairly hot loop, might be worth doing
```
if logrus.GetLogLevel() >= logrus.DebugLevel {
```
to avoid allocating the log entry if debug is disabled.